### PR TITLE
Feature/input box component

### PR DESCRIPTION
**Pull request by llali**
- Added Input box component 
- Added event raising from components to extension

### CODE REVIEW

**Comment by MattIrv** [comment]
It would probably be good for the element to have a css margin set so that it doesn't touch the corners of its container, like the label element that already exists. Not a big concern since we'll probably want to go back and make sure everything is styled consistently after all the elements are done

screen shot 2018-04-19 at 4 07 31 pm

File: src/sql/parts/modelComponents/inputbox.component.ts [hunk 1]
```diff
@@ -0,0 +1,102 @@
...
+import { IComponent, IComponentDescriptor, IModelStore, ComponentEventType } from 'sql/parts/modelComponents/interfaces';
+import { InputBox, IInputOptions } from 'vs/base/browser/ui/inputbox/inputBox';
+import { CommonServiceInterface } from 'sql/services/common/commonServiceInterface.service';
+import { attachInputBoxStyler, attachListStyler } from 'vs/platform/theme/common/styler';
+
+@Component({
+	selector: 'inputBox',
+	template: `
+		<div #input style="width: 100%"></div>
+		<div class="inputBoxComponent" style="position: absolute; height: 100%; width: 100%; margin: 10px 0px 10px 0px;">
```
Comment (kevcunnane) [inline src/sql/parts/modelComponents/inputbox.component.ts]: Quick question - is this class attribute needed? You don't define any CSS file with the properties for it, instead you've put the style information directly into the HTML here. I think you can remove?

File: src/sql/parts/modelComponents/inputbox.component.ts [hunk 2]
```diff
@@ -0,0 +1,102 @@
...
+			<span style="margin-left: 10px; display: inline-block;">
+				<div #input style="width: 100%"></div>
```
Comment (kevcunnane) [inline src/sql/parts/modelComponents/inputbox.component.ts]: Why have 2 divs with `#input` as their selector? The `@ViewChild` reference will only detect the one, right? Shouldn't you have a different name for the 2nd? I don't understand how this works.

File: src/sql/parts/modelComponents/viewBase.ts [hunk 3]
```diff
@@ -91,4 +96,18 @@ export abstract class ViewBase extends AngularDisposable implements IModelView {
 			// TODO add error handling
 		});
 	}
+
+	registerEvent(componentId: string, eventName: string) {
```
Comment (kevcunnane) [inline src/sql/parts/modelComponents/viewBase.ts]: You define `eventName` here but never use it. Will you be addressing this in a future update? Or can you remove it?

File: src/sql/workbench/api/node/extHostModelView.ts [hunk 4]
```diff
@@ -34,27 +35,50 @@ class ModelBuilderImpl implements sqlops.ModelBuilder {
...
+		return this.withEventHandler(new ComponentWrapper(this._proxy, this._handle, ModelComponentTypes.DashboardWebview, id), id);
+	}
+
+	withEventHandler<T extends sqlops.Component>(component: ComponentWrapper, id: string): sqlops.ComponentBuilder<T> {
+		let componentBuilder: ComponentBuilderImpl<T> = new ComponentBuilderImpl<T>(component);
+		this._eventHandlers.set(id, componentBuilder);
+		return componentBuilder;
+	}
+
+	handleEvent(id: string, eventArgs: IComponentEventArgs): void {
```
Comment (kevcunnane) [inline src/sql/workbench/api/node/extHostModelView.ts]: We have lots of IDs. What is this one - a component ID?

File: src/sql/workbench/api/node/extHostModelView.ts [hunk 5]
```diff
@@ -34,27 +35,50 @@ class ModelBuilderImpl implements sqlops.ModelBuilder {
...
 
-class ComponentBuilderImpl<T extends sqlops.Component> implements sqlops.ComponentBuilder<T> {
+interface IWithEventHandler {
+	handleEvent(eventArgs: IComponentEventArgs): void;
+}
+
+class ComponentBuilderImpl<T extends sqlops.Component> implements sqlops.ComponentBuilder<T>, IWithEventHandler {
 
 	constructor(protected _component: ComponentWrapper) {
+		_component.registerEvent();
```
Comment (kevcunnane) [inline src/sql/workbench/api/node/extHostModelView.ts]: Do all components need to register for event handling then?

File: src/sql/workbench/api/node/extHostModelView.ts [hunk 6]
```diff
@@ -301,4 +376,11 @@ export class ExtHostModelView implements ExtHostModelViewShape {
 		this._modelViews.set(handle, view);
 		this._handlers.get(id)(view);
 	}
+
+	$handleEvent(handle: number, id: string, eventArgs: IComponentEventArgs): void {
```
Comment (kevcunnane) [inline src/sql/workbench/api/node/extHostModelView.ts]: Again it's good to define exactly which ID this is. ComponentID?

File: src/sql/parts/modelComponents/inputbox.component.ts (hunk 1 above)
Comment (llali) [reply src/sql/parts/modelComponents/inputbox.component.ts]: I have to figure out how to do the css settings. another PR maybe

---
In reply to: [183179021](https://github.com/Microsoft/sqlopsstudio/pull/1190#discussion_r183179021) [](ancestors = 183179021)

File: src/sql/workbench/api/node/extHostModelView.ts (hunk 5 above)
Comment (llali) [reply src/sql/workbench/api/node/extHostModelView.ts]: I think  all components want to have this by default. I make it optional so the component to register or not 

---
In reply to: [183179896](https://github.com/Microsoft/sqlopsstudio/pull/1190#discussion_r183179896) [](ancestors = 183179896)

File: src/sql/workbench/api/node/extHostModelView.ts (hunk 4 above)
Comment (llali) [reply src/sql/workbench/api/node/extHostModelView.ts]: yes this is the component id

---
In reply to: [183179752](https://github.com/Microsoft/sqlopsstudio/pull/1190#discussion_r183179752) [](ancestors = 183179752)

**Pull request merged**